internal/tracinginterceptor/tracedstream.go

```diff
@@ -0,0 +1,178 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tracinginterceptor
+
+import (
+	"context"
+
+	"github.com/opentracing/opentracing-go"
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/yarpcerrors"
+	"go.uber.org/zap"
+)
+
+// tracedServerStream implements the transport.ServerStream and additional tracing capabilities.
+type tracedServerStream struct {
+	*transport.ServerStream
+	span             opentracing.Span
+	isApplicationErr bool
+	appErrorMeta     *transport.ApplicationErrorMeta
+	log              *zap.Logger
+}
+
+// newTracedServerStream creates a new tracedServerStream with embedded ServerStream.
+func newTracedServerStream(s transport.ServerStream, span opentracing.Span, logger *zap.Logger) *tracedServerStream {
+	return &tracedServerStream{
+		ServerStream: &s,
+		span:         span,
+		log:          logger,
+	}
+}
+
+func (t *tracedServerStream) IsApplicationError() bool {
+	return t.isApplicationErr
+}
+
+func (t *tracedServerStream) ApplicationErrorMeta() *transport.ApplicationErrorMeta {
+	return t.appErrorMeta
+}
+
+func (t *tracedServerStream) SetApplicationError() {
+	t.isApplicationErr = true
+}
+
+func (t *tracedServerStream) SetApplicationErrorMeta(appErrorMeta *transport.ApplicationErrorMeta) {
+	t.appErrorMeta = appErrorMeta
+}
+
+func (t *tracedServerStream) SendMessage(ctx context.Context, msg *transport.StreamMessage) error {
+	err := t.ServerStream.SendMessage(ctx, msg)
+	t.isApplicationErr, t.appErrorMeta = setApplicationErrorDetails(t.span, err)
+	if updateErr := updateSpanWithErrorDetails(t.span, t.isApplicationErr, t.appErrorMeta, err); updateErr != nil {
+		t.log.Error("Failed to update span with error details", zap.Error(updateErr))
+	}
+	return err
+}
+
+func (t *tracedServerStream) ReceiveMessage(ctx context.Context) (*transport.StreamMessage, error) {
+	msg, err := t.ServerStream.ReceiveMessage(ctx)
+	t.isApplicationErr, t.appErrorMeta = setApplicationErrorDetails(t.span, err)
+	if updateErr := updateSpanWithErrorDetails(t.span, t.isApplicationErr, t.appErrorMeta, err); updateErr != nil {
+		t.log.Error("Failed to update span with error details", zap.Error(updateErr))
+	}
+	return msg, err
+}
+
+// tracedClientStream wraps ClientStream with tracing and error metadata.
+type tracedClientStream struct {
+	*transport.ClientStream
+	span             opentracing.Span
+	isApplicationErr bool
+	appErrorMeta     *transport.ApplicationErrorMeta
+	log              *zap.Logger
+}
+
+func newTracedClientStream(cs *transport.ClientStream, span opentracing.Span, logger *zap.Logger) *tracedClientStream {
+	return &tracedClientStream{
+		ClientStream: cs,
+		span:         span,
+		log:          logger,
+	}
+}
+
+func (t *tracedClientStream) IsApplicationError() bool {
```

Comment:
I think the point here is we really don't need `isApplicationErr`, `appErrorMeta` and their relevant setter/getter methods. We can simplify the implementation of `tracedClientStream`:

### Struct Definition
We just need below fields: 
```
type tracedClientStream struct {
	clientStream  *transport.ClientStream
	span              opentracing.Span
	closed           atomic.Bool 
}
```
The `closed` is to avoid double closing a span. [Reference]( https://github.com/yarpc/yarpc-go/blob/dev/transport/grpc/stream.go#L108)

### Struct Methods
You just need to implement the `transport.StreamCloser` interface:
* `Context() context.Context`
* `Request() *StreamRequest`
* `SendMessage(context.Context, *StreamMessage) error`
* `ReceiveMessage(context.Context) (*StreamMessage, error)`
* `Close(context.Context) error`

And you will probably need one more private method for closing the span:
* `closeWithErr(err error) error`
```
func (t *tracedClientStream) Close(ctx context.Context) error {
     return closeWithErr(t.Close(ctx))
}

func (t *tracedClientStream) SendMessage(ctx context.Context, msg *StreamMessage) error{
      if err := t.clientStream.SendMessage(ctx, msg); err != nil {
           return closeWithErr(err)
      }
     return nil
}

func (t *tracedClientStream) ReceiveMessage(ctx context.Context) (*StreamMessage, error) {
      msg,  err := t.clientStream.ReceiveMessage(ctx)
      if err != nil {
         if err ==  io.EOF {
              closeWithErr(nil)
         } else {
              closeWithErr(err)
         }
      }
     return  msg, err
}

func (t *tracedClientStream) closeWithErr(err error) error {
     if !cs.closed.Swap(true) {
		updateSpanWithErrorDetails(span, err != nil, nil, err)
		t.span.Finish()
	}
     return err
}
```